stellar/contract/round/src/internal.rs

```diff
@@ -0,0 +1,802 @@
+            assert!(is_project_in_pair.is_some(), "Project not in pair");
+
+            let pick_result = PickResult {
+                pair_id: picked_pair.pair_id,
+                project_id: picked_pair.voted_project_id,
+            };
+
+            increment_voting_count(env, picked_pair.voted_project_id);
+            picked_pairs.push_back(pick_result);
+        });
+
+        let voting_result = VotingResult {
+            voter: voter.clone(),
+            picks: picked_pairs,
+            voting_time: current_time,
+        };
+
+        add_voting_result(env, voting_result.clone());
+        set_voting_state(env, voter, true);
+        extend_instance(env);
+        log_vote(env, round.id, voting_result);
+    }
+
+    fn get_pair_to_vote(env: &Env) -> Vec<Pair> {
+        let round = read_round_info(env);
+        let pairs = get_random_pairs(env, round.num_picks_per_voter.into());
+        extend_instance(env);
+        pairs
+    }
+
+    fn flag_voter(env: &Env, admin: Address, voter: Address) {
+        admin.require_auth();
+
+        let round = read_round_info(env);
+
+        if round.owner != admin {
+            round
+                .admins
+                .iter()
+                .find(|round_admin| round_admin == &admin)
+                .expect("Only round owner or round admin can flag voter");
+        }
+
+        let is_white_listed = is_white_listed(env, voter.clone());
+        assert!(!is_white_listed, "Voter is white listed");
+
+        let is_black_listed = is_black_listed(env, voter.clone());
+        assert!(!is_black_listed, "Voter already black listed");
+
+        add_to_black_list(env, voter.clone());
+        extend_instance(env);
+        log_update_user_flag(env, round.id, voter, true);
+    }
+
+    fn unflag_voter(env: &Env, admin: Address, voter: Address) {
+        admin.require_auth();
+
+        let round = read_round_info(env);
+
+        if round.owner != admin {
+            round
+                .admins
+                .iter()
+                .find(|round_admin| round_admin == &admin)
+                .expect("Only round owner or round admin can flag voter");
+        }
+
+        let is_black_listed = is_black_listed(env, voter.clone());
+        assert!(is_black_listed, "Voter is not black listed");
+
+        remove_from_black_list(env, voter.clone());
+        extend_instance(env);
+        log_update_user_flag(env, round.id, voter, false);
+    }
+
+    fn calculate_results(env: &Env) -> Vec<ProjectVotingResult> {
+        let results = calculate_voting_results(env);
+        extend_instance(env);
+
+        results
+    }
+
+    fn trigger_payouts(env: &Env, admin: Address) {
+        admin.require_auth();
+
+        let round = read_round_info(env);
+        let current_time = env.ledger().timestamp();
+
+        assert!(
+            round.start_time <= current_time,
+            "Voting period is not started"
+        );
```

Comment:
trigger payouts must be after round.end_time right?

Reply:
yes